piet-coregraphics/src/text.rs

```diff
@@ -783,82 +741,114 @@ impl CoreGraphicsTextLayout {
     }
 
     fn line_number_for_utf8_offset(&self, offset: usize) -> usize {
-        match self.line_offsets.binary_search(&offset) {
+        match self
+            .line_metrics
+            .binary_search_by_key(&offset, |lm| lm.start_offset)
+        {
             Ok(line) => line,
             Err(line) => line.saturating_sub(1),
         }
     }
 
-    /// for each line in a layout, determine its offset in utf8.
-    #[allow(clippy::while_let_on_iterator)]
-    fn rebuild_line_offsets(&mut self) {
-        let lines = self.unwrap_frame().get_lines();
-
-        let utf16_line_offsets = lines.iter().map(|l| {
-            let line = Line::new(&l);
-            let range = line.get_string_range();
-            range.location as usize
-        });
-
-        let mut chars = self.text.chars();
-        let mut cur_16 = 0;
-        let mut cur_8 = 0;
-
-        self.line_offsets = utf16_line_offsets
-            .map(|off_16| {
-                if off_16 == 0 {
-                    return 0;
-                }
-                while let Some(c) = chars.next() {
-                    cur_16 += c.len_utf16();
-                    cur_8 += c.len_utf8();
-                    if cur_16 == off_16 {
-                        return cur_8;
-                    }
-                }
-                panic!("error calculating utf8 offsets");
-            })
-            .collect();
-    }
-
     fn line_range(&self, line: usize) -> Option<(usize, usize)> {
         // no lines: we return the empty string
+        //FIXME: just never have no lines
```

Comment:
Wouldn't this function be slightly cleaner if it returned a Range rather than a pair? Otoh, I wouldn't be surprised if Range created some other problem.

Reply:
I forget exactly the motivation for this originally, and it feels like an "if it ain't broke" situation. This is actually pretty redundant now, since `LineMetric` stores the line range, and we can just use that if we want. I tried to keep this patch minimal.